### ISSUE TITLE
mongodb-2.4.6nb1 (on base64-13.3.0) SMF failure

### ISSUE THREAD

**Issue by nshalman** [TritonDataCenter/pkgsrc-legacy#160]
SMF service in maintenance:

```
[ Dec 24 21:13:18 Leaving maintenance because clear requested. ]
[ Dec 24 21:13:18 Enabled. ]
[ Dec 24 21:13:18 Executing start method ("/opt/local/lib/svc/method/mongodb start"). ]
about to fork child process, waiting until server is ready for connections.
forked process: 14286
Tue Dec 24 21:13:19.021 terminate() called, printing stack (if implemented for platform):
0xecd5f8 0xa85e63 0xfffffd7fabedfd39 0xfffffd7fabedfd83 0xfffffd7fabedffc6 0xfffffd7fabef7ac6 0xfffffd7fabef3c89 0xfffffd7fabeeab95 0xfffffd7fabeebd39 0xfffffd7fabff2770 0xfffffd7fabff4061 0xfffffd7fabfebd39 0xc7f49c 0x1202f63 0xa8589c 
 /opt/local/bin/mongod'_ZN5mongo15printStackTraceERSo+0x28 [0xecd5f8]
 /opt/local/bin/mongod'_ZN5mongo11myterminateEv+0x43 [0xa85e63]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZN10__cxxabiv111__terminateEPFvvE+0x9 [0xfffffd7fabedfd39]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'0xdfd83 [0xfffffd7fabedfd83]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'0xdffc6 [0xfffffd7fabedffc6]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZSt21__throw_runtime_errorPKc+0x56 [0xfffffd7fabef7ac6]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'0xf3c89 [0xfffffd7fabef3c89]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZNSt6locale5_ImplC2EPKcm+0x45 [0xfffffd7fabeeab95]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZNSt6localeC2EPKc+0x7e9 [0xfffffd7fabeebd39]
 /opt/local/lib/libboost_filesystem.so.1.53.0'_ZN5boost10filesystem4path7codecvtEv+0x50 [0xfffffd7fabff2770]
 /opt/local/lib/libboost_filesystem.so.1.53.0'_ZNK5boost10filesystem4path14root_directoryEv+0x101 [0xfffffd7fabff4061]
 /opt/local/lib/libboost_filesystem.so.1.53.0'_ZN5boost10filesystem8absoluteERKNS0_4pathES3_+0x29 [0xfffffd7fabfebd39]
 /opt/local/bin/mongod'_ZN5mongo27initializeServerGlobalStateEb+0xec [0xc7f49c]
 /opt/local/bin/mongod'main+0x243 [0x1202f63]
 /opt/local/bin/mongod'_start+0x6c [0xa8589c]
Tue Dec 24 21:13:19.022 Got signal: 6 (Abort).

Tue Dec 24 21:13:19.022 Backtrace:
0xecd5f8 0xa862eb 0xfffffd7fff17ced6 0xfffffd7fff16fd2b 0xfffffd7fff183ffa 0xfffffd7fff11d140 0xfffffd7fff0f7930 0xa85e68 0xfffffd7fabedfd39 0xfffffd7fabedfd83 0xfffffd7fabedffc6 0xfffffd7fabef7ac6 0xfffffd7fabef3c89 0xfffffd7fabeeab95 0xfffffd7fabeebd39 0xfffffd7fabff2770 0xfffffd7fabff4061 0xfffffd7fabfebd39 0xc7f49c 0x1202f63 
 /opt/local/bin/mongod'_ZN5mongo15printStackTraceERSo+0x28 [0xecd5f8]
 /opt/local/bin/mongod'_ZN5mongo10abruptQuitEi+0x11b [0xa862eb]
 /lib/amd64/libc.so.1'__sighndlr+0x6 [0xfffffd7fff17ced6]
 /lib/amd64/libc.so.1'call_user_handler+0x1db [0xfffffd7fff16fd2b]
 /lib/amd64/libc.so.1'_lwp_kill+0xa [0xfffffd7fff183ffa]
 /lib/amd64/libc.so.1'raise+0x20 [0xfffffd7fff11d140]
 /lib/amd64/libc.so.1'abort+0x60 [0xfffffd7fff0f7930]
 /opt/local/bin/mongod'_ZN5mongo11myterminateEv+0x48 [0xa85e68]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZN10__cxxabiv111__terminateEPFvvE+0x9 [0xfffffd7fabedfd39]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'0xdfd83 [0xfffffd7fabedfd83]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'0xdffc6 [0xfffffd7fabedffc6]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZSt21__throw_runtime_errorPKc+0x56 [0xfffffd7fabef7ac6]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'0xf3c89 [0xfffffd7fabef3c89]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZNSt6locale5_ImplC2EPKcm+0x45 [0xfffffd7fabeeab95]
 /opt/local/gcc47/x86_64-sun-solaris2.11/lib/amd64/libstdc++.so.6.0.17'_ZNSt6localeC2EPKc+0x7e9 [0xfffffd7fabeebd39]
 /opt/local/lib/libboost_filesystem.so.1.53.0'_ZN5boost10filesystem4path7codecvtEv+0x50 [0xfffffd7fabff2770]
 /opt/local/lib/libboost_filesystem.so.1.53.0'_ZNK5boost10filesystem4path14root_directoryEv+0x101 [0xfffffd7fabff4061]
 /opt/local/lib/libboost_filesystem.so.1.53.0'_ZN5boost10filesystem8absoluteERKNS0_4pathES3_+0x29 [0xfffffd7fabfebd39]
 /opt/local/bin/mongod'_ZN5mongo27initializeServerGlobalStateEb+0xec [0xc7f49c]
 /opt/local/bin/mongod'main+0x243 [0x1202f63]

ERROR: child process failed, exited with error number 14
[ Dec 24 21:13:19 Method "start" exited with status 14. ]
```

But...

```
# su - mongodb -c "/opt/local/lib/svc/method/mongodb start"
about to fork child process, waiting until server is ready for connections.
forked process: 14294
all output going to: /var/log/mongodb/mongodb.log
child process started successfully, parent exiting
```

**Comment by nshalman** [comment]
(Hacky?) solution:
Set LC_ALL=C and specify that this should be a contract.
Also there's a nit in `/opt/local/lib/svc/method/mongodb` of an unexpanded `@SMF_HOME@`

Diffs:

``` diff
--- /opt/local/lib/svc/method/mongodb
+++ mongodb
@@ -5,7 +5,7 @@
 . /lib/svc/share/smf_include.sh

 PATH="/opt/local/bin:/opt/local/gnu/bin:/opt/local/sbin:/usr/bin:/usr/sbin";
-DATADIR="@SMF_HOME@";
+DATADIR="/var/mongodb";
 MONGO_CONF="/opt/local/etc/mongodb.conf";
 MONGO_PID="${DATADIR}/mongodb.pid";

--- /opt/local/lib/svc/manifest/mongodb.xml
+++ mongodb.xml
@@ -17,11 +17,15 @@
       <method_credential group='mongodb' user='mongodb'/>
         <method_environment>
           <envvar name='PATH' value='/opt/local/bin:/opt/local/gnu/bin:/opt/local/sbin:/usr/bin:/usr/sbin'/>
+          <envvar name='LC_ALL' value='C'/>
         </method_environment>
     </method_context>
     <exec_method name='start' type='method' exec='/opt/local/lib/svc/method/mongodb start' timeout_seconds='18446744073709551615'/>
     <exec_method name='stop' type='method' exec='/opt/local/lib/svc/method/mongodb stop' timeout_seconds='18446744073709551615'/>
     <exec_method name='restart' type='method' exec='/opt/local/lib/svc/method/mongodb restart' timeout_seconds='18446744073709551615'/>
+    <property_group name='startd' type='framework'>
+          <propval name='duration' type='astring' value='contract' />
+    </property_group>
       <property_group name='replication' type='application'>
        <propval name='name' type='astring' value=''/>
        <propval name='key' type='astring' value=''/>
```

**Comment by mamash** [comment]
The real problem is an image (base/base64) regression introduced in 13.3.0. The default locale for any image should be C, not en_US.UTF-8 (set in /etc/default/init). We'll fix this in the next image revision.

The unexpanded SMF_HOME placeholder is a valid point, we'll fix that in the next package rebuild.

**Comment by nshalman** [comment]
What about telling SMF to run it in a contract (since the start method forks it into the background)?

**Comment by mamash** [comment]
Contract is the default value for startd/duration.

**Comment by mamash** [comment]
This was fixed in base 13.3.1 released earlier, and the mongodb package was rebuild to take care of the other problem.
